worlds/AutoWorld.py

```diff
@@ -365,133 +364,311 @@ def stage_assert_generate(cls, multiworld: "MultiWorld") -> None:
+        options.
+
+        :param spoiler_handle: The spoiler file for this generation. The file is already open when this method is
+        called.
         """
-        pass
 
     def write_spoiler(self, spoiler_handle: TextIO) -> None:
         """
-        Write to the spoiler "middle", this is after the per-player options and before locations,
-        meant for useful or interesting info.
+        Write to the spoiler "middle", this is after the per-player options and topology information, but before
+        locations information. Meant for useful or interesting info.
+
+        :param spoiler_handle: The spoiler file for this generation. The file is already open when this method is
+        called.
+        """
+
+    @classmethod
+    def stage_write_spoiler(cls, multiworld: "MultiWorld", spoiler_handle: TextIO) -> None:
+        """
+        Class level stage of write_spoiler. Gets run once per world type per multiworld after all the instanced calls.
+
+        :param multiworld: The multiworld object for this generation.
+        :param spoiler_handle: The spoiler file for this generation. The file is already open when this method is
+        called.
         """
-        pass
 
     def write_spoiler_end(self, spoiler_handle: TextIO) -> None:
-        """Write to the end of the spoiler"""
-        pass
+        """
+        Write to the end of the spoiler. Called after all other spoiler information has been written.
+
+        :param spoiler_handle: The spoiler file for this generation. The file is already open when this method is
+        called.
+        """
+
+    @classmethod
+    def stage_write_spoiler_end(cls, multiworld: "MultiWorld", spoiler_handle: TextIO) -> None:
+        """
+        Class level stage of write_spoiler_end. Gets run once per world type per multiworld after all the instanced
+        calls.
+
+        :param multiworld: The multiworld object for this generation.
+        :param spoiler_handle: The spoiler file for this generation. The file is already open when this method is
+        called.
+        """
 
     # end of ordered Main.py calls
 
     def create_item(self, name: str) -> "Item":
         """
         Create an item for this world type and player.
-        Warning: this may be called with self.world = None, for example by MultiServer
+        Warning: this may be called with self.world = None, for example by MultiServer.
+
+        :param name: The name of the item to be created.
+        :return: The created item.
         """
         raise NotImplementedError
 
     def get_filler_item_name(self) -> str:
-        """Called when the item pool needs to be filled with additional items to match location count."""
+        """
+        Called when the item pool needs to be filled with additional items to match location count.
+
+        :return: The name of the item to be created.
```

Comment:
This name then gets passed to `create_item`, right? Is that worth mentioning here?